accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -107,6 +107,44 @@ The internal representation of types is extended with a type `T*` for every type
 `T` to represent legacy pre-NNBD types.  This is discussed further in the legacy
 library section below.
 
+### Subtyping
+
+We modify the subtyping rules to account for nullability and legacy types as
+specified
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md).
+We write `S <: T` to mean that the type `S` is a subtype of `T` according to the
+rules specified there.
+
+
+We define `LEGACY_SUBTYPE(S, T)` to be true iff `S` would be a subtype of `T`
```

Comment:
I think so, but I haven't thought deeply about it.  Is this a useful property for you?

Reply:
It looked to me like it might be equivalent, and implementing it as a composition of two other operations, which we have anyway, is easier than patching the NNBD version of `SUBTYPE` to do similar transformations to erase nullability.

Reply:
These should be equivalent. I was under the impression that you already had the two subtyping algorithms implemented though?

Reply:
There is no second implementation, my understanding was that the `NNBD_SUBTYPE` works for legacy types. So, analyzer uses it inside of both legacy and opted-in libraries, for all types.

Reply:
I guess then you must be already using erasure then, since the analyzer accepts this example:

```dart
// opted in
class A {
  int? foo(int? x) {}
}

class B {
  int foo(int x) {}
}

// opted out
void test() {
  new B().foo(new A().foo(3));
}
```

The assignment of `int?` to `int` in `test` is invalid according to NNBD_SUBTYPE, so I assume you are already implementing this as the composition of NNBD_SUBTYPE and LEGACY_ERASURE?

Reply:
Yes, we already implement `LEGACY_ERASURE` to some extent. I did not think about any Never / Null conversion, and discarding `required`, will update it.

We erase nullabilities after lookup, so `B().foo` is `int* Function(int*)`, and `A().foo` is also `int* Function(int*)`, so all types are compatible. There was name for this, but it looks that yes, we de-facto implemented subtype checks as such composition. We need elements for all identifiers, so for `A().foo` and `B().foo` we record that elements (`MethodMember`) that they reference `A::foo` and `B::foo` and their `type`s are after `LEGACY_ERASURE`.